File: Numerical Computing Library/NumericalLibrary-ver2/library/MatrixV1_Python/matrixv1/operations/multiply.py
from matrixv1.core.matrix import Matrix


def multiply(A: Matrix, B: Matrix) -> Matrix:
    if A.cols != B.rows:
        raise ValueError(
            f"Invalid dimensions for multiply: {A.rows}x{A.cols} * {B.rows}x{B.cols}"
        )
    result = Matrix(A.rows, B.cols)
    for i in range(A.rows):
        for j in range(B.cols):
            total = sum(A.get(i, k) * B.get(k, j) for k in range(A.cols))
            result.set(i, j, total)
    return result

File: Numerical Computing Library/NumericalLibrary-ver2/library/MatrixV1/src/operations/Multiply.cpp
#include "../../include/operations/Multiply.hpp"
#include <stdexcept>

Matrix Multiply::compute(const Matrix& A, const Matrix& B) {
    if(A.getCols() != B.getRows())
        throw std::invalid_argument("Invalid dimensions");

    Matrix result(A.getRows(), B.getCols());

    for(int i = 0; i < A.getRows(); i++) {
        for(int j = 0; j < B.getCols(); j++) {
            double sum = 0;
            for(int k = 0; k < A.getCols(); k++) {
                sum += A.get(i,k) * B.get(k,j);
            }
            result.set(i, j, sum);
        }
    }

    return result;
}